design/0005-protocol-detection.md

```diff
@@ -0,0 +1,308 @@
+
+### Requirements
+
+- Preserve the existing port-oriented proxy bypass functionality (`skip-ports`).
+- Support mTLS & TCP telemetry for arbitrary TCP protocols, including server-first protocols.
+- Support both half-meshed and full-meshed client-server relationships.
+- Clients do not need to document/configure the endpoints to which they connect.
+
+### Design
+
+This proposal assumes that Linkerd's outbound proxy implements discovery for
+mTLS to arbitrary TCP endpoints.
+
+The destination and injection controllers are modified to honor a new
+annotation, which may only be set on ``Pod`` resources:
+
+```yaml
+config.linkerd.io/proxy-opaque-tcp: <port-list>
+```
+
+Where `port-list` is a comma-separated list of port numbers, port ranges,
+and/or port names.
+
+For example:
+
+```yaml
+---
+apiVersion: apps/v1beta2
+kind: Deployment
+metadata:
+  name: mysql
+  labels:
+    app: mysql
+spec:
+  selector:
+    matchLabels:
+      app: mysql
+  strategy:
+    type: Recreate
+  template:
+    metadata:
+      labels:
+        app: mysql
+      annotations:
+        config.linkerd.io/proxy-opaque-tcp: mysql
+    spec:
+      containers:
+        - name: mysql
+          image: mysql:5.6
+          ports:
+            - containerPort: 3306
+              name: mysql
+```
+
+### Injection Controller
+
+The injector encodes this port list in a proxy environment configuration,
+`LINKERD2_PROXY_INBOUND_OPAQUE_PORTS`. When an inbound proxy receives any
+inbound traffic on these ports, the connection is forwarded blindly to the
+application (without mTLS, etc).
+
+### Destination Controller
+
+The destination controller uses this annotation when assisting the outbound
+proxy's protocol discovery via the new `Destination.Strategy` API endpoint
+(to be added as a part of TCP work), this annotation instructs the proxy to
+skip protocol detection on this stream and to forward it opaquely. This
+should be honored for both meshed and un-meshed workloads.
+
+When resolving a `Service` address, the union of all values from all pods in
+that service is considered.
+
+When communicating with a meshed endpoint, the destination response includes
+a protocol upgrade hint including the remote inbound proxy's port number and
+its identity (for use with mTLS).
+
+The outbound proxy then connects to the remote inbound port _directly_ (and
+not the original destination port). The endpoint's application port (as
+provided by the discovery response and **not** the `SO_ORIG_DST` lookup),
+must be preserved and transmitted to the remote proxy so that it can route it
+properly. This is transmitted via a new _Linkerd Connection Header_.
+
+#### Multi-Cluster Service Mirror Controller
+
+The service-mirror creates local endpoints that reflect the opaqueness, as
+configured in the target cluster. The multi-cluster gateway then uses the
+connection header to route the connection through the gateway.
+
+### Proxy handling of inbound-targeted connections
```

Comment:
Just so I make sure I understand correctly... The search for ` _Linkerd Connection Header_`  is happening only when the inbound connection is happening on the proxy inbound port directly?

Reply:
The RFC shouldn't specify that, imo. I'm only trying to define the behavior for what happens when a connection targets the inbound port. It's an implementation detail as to whether we attempt this on other connections...